### ISSUE TITLE
Top Gear Rally bad audio

### ISSUE THREAD

**Issue by fzurita** [mupen64plus/mupen64plus-core#200]
I've been trying to figure out what's wrong with audio in this game, but I'm not having much luck. The one thing that seems to be special about this game is that audio comes in 16384 byte chunks to the audio plugin when `AiLenChanged` is called, normally I see around 2KB chunks.

I dumped these audio samples to a text file as they come into the audio plugin just to see what the values look like, I don't see anything unreasonable, but when audio is played back from this data, it doesn't sound right.

Attached are audio samples for that game in text format. Each row is a byte array that is read when AiLenChanged is called.

[audio_output.zip](https://github.com/mupen64plus/mupen64plus-core/files/560707/audio_output.zip)

At this point, I think that the audio coming from the core is not right. I have tried switching RSP between CXD4 and RSP-HLE with no effect.

**Comment by Gillou68310** [comment]
Did you tried with cxd4 plugin?

**Comment by fzurita** [comment]
Yeah, no improvement.

**Comment by Gillou68310** [comment]
Sorry I just saw you already mentioned it in the first post.

**Comment by fzurita** [comment]
No problem

**Comment by Gillou68310** [comment]
What about pj64?

**Comment by fzurita** [comment]
Very similar issue with default settings.

**Comment by Gillou68310** [comment]
Ok this will probably be hard to fix then. We have no working case to compare with.

**Comment by mudlord** [comment]
You mean you need working audio samples of the background music and stuff?

Top Gear Rally's music is tracked. You can get it from:
http://ftp.modland.com/pub/modules/Fasttracker%202/Barry%20Leitch/

**Comment by loganmc10** [comment]
Now that World Driver Chamionship and Stunt Racer are working, they seem to have the same issue. Really bad audio, and the audio comes in 16384 byte chunks. The audio doesn't sound great in Project64 either.

**Comment by loganmc10** [comment]
I've been looking into this, and I believe I know the cause. I'm going to write out all the details in case anyone in the future finds this interesting.

When a game wants to play some audio, it writes the address of the audio buffer to AI_DRAM_ADDR_REG, and it writes the length of the audio buffer to AI_LEN_REG. Writing to AI_LEN_REG triggers a DMA that starts audio playing.

The way we emulate this is: When AI_LEN_REG is written (DMA starts), we copy the contents of the audio buffer to the audio plugin and play it.

For a majority of games this works well, but Boss Games are unique.

Firstly, most games have multiple places in memory for the audio buffers, so when one buffer is complete, it updates AI_DRAM_ADDR_REG with the next location, which has the next data, and starts a new DMA.

Boss Games have only 1 memory location for audio (AI_DRAM_ADDR_REG never changes). This means that when a DMA is complete and a new one starts, it just reads the same memory location.

In order to update the audio buffer, the games are modifying it's contents **while the DMA is happening**.

To prove this point, I took CRC's of the audio buffer at the beginning and end of the DMA, as well as anytime AI_LEN_REG is read (reading AI_LEN_REG tells the game how much of the audio buffer is left to play.

An example from Super Mario 64:
```
start dma 8a4ad553
read ai_len 8a4ad553
end dma 8a4ad553

start dma e4516a6f
read ai_len e4516a6f
end dma e4516a6f

start dma 1d7428e5
read ai_len 1d7428e5
end dma 1d7428e5
```

Golden Eye:
```
start dma 56bfbf75
read ai_len 56bfbf75
read ai_len 56bfbf75
end dma 56bfbf75

start dma 9fbd1124
read ai_len 9fbd1124
read ai_len 9fbd1124
end dma 9fbd1124

start dma fb570a52
read ai_len fb570a52
read ai_len fb570a52
end dma fb570a52
```

And now, Top Gear Really:
```
start dma 8f888956
read ai_len 8f888956
read ai_len f8e2d4b1
read ai_len 90f5bc96
read ai_len d6c4487a
read ai_len 22df6904
read ai_len 88735031
read ai_len 888366e4
read ai_len e8b1ae9
read ai_len 30ce1a44
read ai_len 8c69b423
read ai_len bb92378c
read ai_len aacd07f2
read ai_len 7621feb
read ai_len b65b000d
read ai_len 5e8d9256
read ai_len a0f9e3ff
read ai_len a76f2a74
read ai_len 5e570b18
read ai_len 7ea2cb43
end dma 5cab8dfc

start dma 5cab8dfc
read ai_len 5cab8dfc
read ai_len 5cb567d0
read ai_len 3afd65f5
read ai_len abda358e
read ai_len 459eb0e6
read ai_len 8dc6b49
read ai_len bbdd89c9
read ai_len 828de352
read ai_len ff0c5e22
read ai_len 2e2a8ecf
read ai_len 2c8e7db1
read ai_len 58c83921
read ai_len c80d862
read ai_len c97bf31f
read ai_len 6157f857
read ai_len d0bfecda
read ai_len d738f4ee
read ai_len 101d5a45
read ai_len a23623cd
end dma 905327f9
```

You see that in Mario, Golden Eye, and most other games, the contents of the audio buffer remain constant throughout the whole DMA, so when we copy the entire contents of the audio buffer at the beginning of the DMA, that's fine, since it's the same the whole way through.

With Boss Games, they are updating the audio buffer while the DMA is happening, so when we copy the entire contents of the audio buffer at the beginning of the DMA and send it to the audio plugin, we aren't sending the correct data. Copying it at the end of the DMA is no good either, you still aren't capturing how it happened on the N64. If we were "cycle accurate", we would copy a little bit from the audio buffer every cycle (not the whole thing at the beginning or end), so this wouldn't be an issue.

I haven't fully tested a solution yet, but as you can see, they read AI_LEN_REG a lot more than other games. I believe these games are reading this value to determine where it currently is in the DMA, and that lets them determine what area of the audio buffer they can update.

So basically we need to emulate this "streaming buffer update". I think this will involve doing a memcpy of the audio buffer piece by piece when AI_LEN_REG is read, and then piecing together the proper audio data at the end of the DMA, and then sending that to the audio plugin. This will introduce a little bit of audio latency, since the audio data is being sent at the end of the DMA, so it will probably be a game specific thing for games that stream data like this.

In the coming week I'll probably test this out a bit, I'm not sure how far I'll get.

**Comment by loganmc10** [comment]
I just tested the latest mupen64plus AE and it no longer sounds like the audio was processed by a garburator, I think this is solved

**Comment by fzurita** [comment]
Cool :)

**Comment by fzurita** [comment]
Confirmed, it sounds great now. Good work.

**Comment by loganmc10** [comment]
Thanks, these Boss Games (Top Gear, World Driver, Twisted Edge Snowboarding..) still have this weird issue where the audio will go berserk (static, strange sounds) every now and then. Top Gear actually doesn't suffer from it too bad, but the other games do.

Messing with the Scanline/Op timing seems to make a difference, so I assume it's just the case of the audio getting out of sync or something. Still a lot better than it used to be at least.
